File: merton_model.py
# Importa las bibliotecas necesarias
from scipy.optimize import fmin  # Para optimización
import numpy as np               # Para operaciones numéricas
import scipy.stats as si         # Para estadísticas

# Define una función NORMSDIST que calcula la función de distribución acumulativa estándar
def NORMSDIST(x):
    NORMSDIST = si.norm.cdf(x, 0.0, 1.0)
    return(NORMSDIST)

# Define la función MertonEquity para calcular el valor de las acciones
def MertonEquity(V, K, r, T, σV):
    d1 = np.log(V / (K * np.exp(-r * T))) / (σV * np.sqrt(T)) + 0.5 * σV * np.sqrt(T)
    d2 = d1 - σV * np.sqrt(T)
    MertonEquity = V * NORMSDIST(d1) - K * np.exp(-r * T) * NORMSDIST(d2)
    return (MertonEquity)

# Define la función MertonDebt para calcular el valor de la deuda
def MertonDebt(V, K, r, T, σV):
    d1 = np.log(V / (K * np.exp(-r * T))) / (σV * np.sqrt(T)) + 0.5 * σV * np.sqrt(T)
    d2 = d1 - σV * np.sqrt(T)
    MertonDebt = K * np.exp(-r * T) * NORMSDIST(d2) + V * (1 - NORMSDIST(d1))
    return (MertonDebt)

# Definición de variables
S = 20       # Valor de mercado de las acciones
K = 99.46    # Valor nominal de la deuda
r = 0.1      # Tasa de interés libre de riesgo
T = 1        # Plazo en años
σS = 0.4     # Volatilidad del valor de mercado de las acciones

# Define la función ImpliedMerton para optimizar los parámetros V y σV
def ImpliedMerton(c):
    d1 = np.log(c[0] / (K * np.exp(-r * T))) / (c[1] * np.sqrt(T)) + 0.5 * c[1] * np.sqrt(T)
    d2 = d1 - c[1] * np.sqrt(T)
    f1 = (c[0] * NORMSDIST(d1) - K * np.exp(-r * T) * NORMSDIST(d2)) - S
    f2 = σS * S - NORMSDIST(d1) * c[1] * c[0]
    val = f1 ** 2 + f2 ** 2
    print("[V, σV]=", c, ", Object Function Value:", val)
    return (val)

# Utiliza la función fmin de SciPy para encontrar los valores óptimos de V y σV que minimizan la función objetivo
c = fmin(ImpliedMerton, [100, 0.1])

File: merton_model.R
# Cargar las bibliotecas necesarias
library(quantmod)

# Definir los parámetros del modelo de Merton
S0 <- 100       # Precio actual de la acción
K <- 105        # Precio de ejercicio de la opción
r <- 0.05       # Tasa de interés libre de riesgo
T <- 1          # Plazo en años
sigma_S <- 0.2  # Volatilidad del activo subyacente

# Generar una secuencia de tiempo para simular el precio de la acción
n <- 252        # Número de días de negociación
dt <- T / n     # Incremento de tiempo
t <- seq(0, T, by = dt)
N <- length(t)

# Simulación de precios usando el modelo de Merton
set.seed(123)   # Establecer una semilla para la reproducibilidad
price_path <- numeric(N)
price_path[1] <- S0

for (i in 2:N) {
  dW <- rnorm(1, mean = 0, sd = sqrt(dt))
  price_path[i] <- price_path[i - 1] * exp((r - 0.5 * sigma_S^2) * dt + sigma_S * dW)
}

# Calcular el valor de la opción de compra (call) usando la fórmula de Merton
d1 <- (log(S0 / K) + (r + 0.5 * sigma_S^2) * T) / (sigma_S * sqrt(T))
d2 <- d1 - sigma_S * sqrt(T)
call_price <- S0 * pnorm(d1) - K * exp(-r * T) * pnorm(d2)  # PARA OPCION DE COMPRA
put_price <- K * exp(-r * T) * pnorm(-d2) - S0 * pnorm(-d1) # PARA OPCION DE VENTA


# Resultados
cat("Precio de la acción en el plazo T:", price_path[N], "\n")
cat("Valor de la opción de compra (call):", call_price, "\n")
cat("Vaor de la opción de venta (put):", put_price)
